### PR TITLE
Account for alignment on `is_plain` calculations

### PR DESCRIPTION
**Pull request by MiguelCompany**
This shall address https://github.com/ros2/rmw_fastrtps/issues/715 by taking into consideration whether the alignment generated by the compiler is compatible with the one required by the CDR serialization.

### CODE REVIEW

**Comment by clalancette** [comment]
Preliminary CI with this and https://github.com/ros2/system_tests/pull/530 and https://github.com/ros2/rmw_fastrtps/pull/716

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19590/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14101/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20314/)

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 1]
```diff
@@ -609,7 +615,23 @@ if isinstance(type_, AbstractNestedType):
...
+  if (is_plain) {
+    // All members are plain, and type is not empty.
+    // We still need to check that the in-memory alignment
+    // is the same as the CDR mandated alignment.
+    using DataType = @('__'.join([package_name] + list(interface_path.parents[0].parts) + [message.structure.namespaced_type.name]));
+    DataType * data;
+    is_plain =
+      (
+        offsetof(DataType, @(last_member_name_) ) +
+        sizeof(decltype(data->@(last_member_name_) ) )
```
Comment (clalancette) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:629]: If the last member in the message was a nested message (the `else` clause above), *and* it was a plain type itself, then I think this sizeof clause might be incorrect, right?  That is, in that case we this should be:

`offsetof(DataType, @(last_member_name_) ) + last_message_size`

where `last_message_size` is what was returned from the recursive call to `max_serialized_size`.  Does that make sense?

**Review by fujitatomoya — COMMENTED** [review]
@MiguelCompany i would like to check my understanding here.

so the problem happens for both `intra` and `inter` process communication, via `rclcpp` right. (but with `rclpy`, that is why we do not see the problem with `ros2 topic echo`?) because the root cause is the mismatch between CDR representation and in-memory data alignment?

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em (hunk 1 above)
Comment (clalancette) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:629]: (separately, uncrustify is unhappy about the indentation here, but we can wait on that until we answer the question above)

Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: @clalancette > If the last member in the message was a nested message (the `else` clause above), _and_ it was a plain type itself, then I think this sizeof clause might be incorrect, right? That is, in that case we this should be:
> 
> `offsetof(DataType, @(last_member_name_) ) + last_message_size`
> 
> where `last_message_size` is what was returned from the recursive call to `max_serialized_size`. Does that make sense?

I have been doing some checks with different nested types, and I got to the conclusion that you are right.

I implemented this behavior in 71b1a0fd6bd5130f711ce687644790e7c04f1e0f

Comment (MiguelCompany) [reply rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em]: > (separately, uncrustify is unhappy about the indentation here, but we can wait on that until we answer the question above)

Fixed in d9ec07e809f7649dee67556d441114ec61f90ecd

**Comment by MiguelCompany** [comment]
> @MiguelCompany i would like to check my understanding here.
> 
> so the problem happens for both `intra` and `inter` process communication, via `rclcpp` right. (but with `rclpy`, that is why we do not see the problem with `ros2 topic echo`?) 

This is reproduced when the sample is accessed via `rmw_take_loaned_message`. I don't know whether rclpy is using it or not.

> the root cause is the mismatch between CDR representation and in-memory data alignment?
The root cause is mismatch between CDR representation and in-memory representation due to different alignment of fields.

So what we are doing with this (and related) PR is disallowing the loans (marking the type as not plain) on those types where there is a representation mismatch.

**Comment by fujitatomoya** [comment]
thanks!

just a comment.

> This is reproduced when the sample is accessed via rmw_take_loaned_message. I don't know whether rclpy is using it or not.

`rclpy` doe not have any interfaces to `rcl/rmw_take_loaned_*`, so we do not have problem. original problem comes **without** `LoanedMessage` class at all on the publisher, this is because subscriber consider `can_loan_messages` as true

https://github.com/ros2/rmw_fastrtps/blob/ad4550adea4f65fdc126423f2fec13f441c6bbaa/rmw_fastrtps_shared_cpp/src/rmw_take.cpp#L565

and then call `rcl_subscription_can_loan_messages` via `rclcpp`.

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 2]
```diff
@@ -581,35 +589,60 @@ if isinstance(type_, AbstractNestedType):
...
       eprosima::fastcdr::Cdr::alignment(current_alignment, sizeof(long double));
 @[    end if]@
 @[  else]
+    last_member_size = 0;
     for (size_t index = 0; index < array_size; ++index) {
       bool inner_full_bounded;
       bool inner_is_plain;
-      current_alignment +=
+      size_t inner_size;
+      inner_size =
```
Comment (clalancette) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:617]:
```suggestion
      inner_size =
```

File: rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em [hunk 3]
```diff
@@ -581,35 +589,60 @@ if isinstance(type_, AbstractNestedType):
...
 
-  return current_alignment - initial_alignment;
+  size_t ret_val = current_alignment - initial_alignment;
+@[if last_member_name_ is not None]@
+  if (is_plain) {
+    // All members are plain, and type is not empty.
+    // We still need to check that the in-memory alignment
+    // is the same as the CDR mandated alignment.
+    using DataType = @('__'.join([package_name] + list(interface_path.parents[0].parts) + [message.structure.namespaced_type.name]));
+    DataType * data;
```
Comment (clalancette) [inline rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em:636]:
```suggestion
```

(it is unused now)

File: rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em [hunk 4]
```diff
@@ -428,35 +436,59 @@ if isinstance(type_, AbstractNestedType):
...
 
-  return current_alignment - initial_alignment;
+  size_t ret_val = current_alignment - initial_alignment;
+@[if last_member_name_ is not None]@
+  if (is_plain) {
+    // All members are plain, and type is not empty.
+    // We still need to check that the in-memory alignment
+    // is the same as the CDR mandated alignment.
+    using DataType = @('::'.join([package_name] + list(interface_path.parents[0].parts) + [message.structure.namespaced_type.name]));
+    DataType * data;
```
Comment (clalancette) [inline rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em:482]:
```suggestion
```

**Review by clalancette — CHANGES_REQUESTED** [review]
Three more small changes, then I think this will be ready for CI again.

**Comment by MiguelCompany** [comment]
@clalancette Applied your suggestions on 0de64df90ddad1ceab2c7dafd4a6e0a8c4a33f91

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19592/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14103/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20316/)

**Comment by fujitatomoya** [comment]
@Mergifyio backport iron humble

**Comment by fujitatomoya** [comment]
@clalancette since i do not have write permission on this repo, can you do me a favor?

**Comment by clalancette** [comment]
@Mergifyio backport iron humble

**Pull request merged**